File: python/x402/http/facilitator_client.py
"""HTTP-based facilitator client for x402 protocol.

Provides both async (HTTPFacilitatorClient) and sync (HTTPFacilitatorClientSync)
implementations for communicating with remote facilitator services.
"""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, Any, TypeVar

from pydantic import ValidationError

from ..schemas import (
    PaymentPayload,
    PaymentRequirements,
    SettleResponse,
    SupportedResponse,
    VerifyResponse,
)
from ..schemas.v1 import PaymentPayloadV1, PaymentRequirementsV1
from .extension_responses import extract_extension_responses_header, log_extension_responses
from .facilitator_client_base import (
    AuthHeaders,
    AuthProvider,
    CreateHeadersAuthProvider,
    FacilitatorClient,
    FacilitatorClientSync,
    FacilitatorConfig,
    FacilitatorResponseError,
    HTTPFacilitatorClientBase,
)
from .utils import (
    ResponseBodyTooLargeError,
    aread_limited_body,
    read_limited_body,
)

if TYPE_CHECKING:
    import httpx

# Re-export for external use
__all__ = [
    "HTTPFacilitatorClient",
    "HTTPFacilitatorClientSync",
    "FacilitatorConfig",
    "FacilitatorResponseError",
    "FacilitatorClient",
    "FacilitatorClientSync",
    "AuthProvider",
    "AuthHeaders",
    "CreateHeadersAuthProvider",
]

_ResponseModelT = TypeVar(
    "_ResponseModelT",
    VerifyResponse,
    SettleResponse,
    SupportedResponse,
)
_ResponseWithSidechannelT = TypeVar("_ResponseWithSidechannelT", VerifyResponse, SettleResponse)


def _response_excerpt(body: bytes, limit: int = 200) -> str:
    """Build a compact response preview for parse errors."""
    text = body.decode("utf-8", errors="replace").strip()
    if not text:
        return "<empty response>"

    compact = " ".join(text.split())
    if len(compact) <= limit:
        return compact
    return f"{compact[: limit - 3]}..."


def _parse_facilitator_response(
    body: bytes,
    model_cls: type[_ResponseModelT],
    operation: str,
) -> _ResponseModelT:
    """Parse facilitator JSON into a validated response model."""
    try:
        response_data = json.loads(body)
    except (json.JSONDecodeError, ValueError, TypeError) as exc:
        raise FacilitatorResponseError(
            f"Facilitator {operation} returned invalid JSON: {_response_excerpt(body)}"
        ) from exc

    try:
        return model_cls.model_validate(response_data)
    except (ValidationError, ValueError, TypeError) as exc:
        raise FacilitatorResponseError(
            f"Facilitator {operation} returned invalid data: {_response_excerpt(body)}"
        ) from exc


def _decode_error_body(body: bytes) -> str:
    return body.decode("utf-8", errors="replace")


def _read_sync_response_body(response: Any) -> bytes:
    try:
        return read_limited_body(response.iter_bytes())
    except ResponseBodyTooLargeError:
        raise
    except Exception as err:
        raise ValueError("failed to read response body") from err
    finally:
        response.close()


async def _read_async_response_body(response: Any) -> bytes:
    try:
        return await aread_limited_body(response.aiter_bytes())
    except ResponseBodyTooLargeError:
        raise
    except Exception as err:
        raise ValueError("failed to read response body") from err
    finally:
        await response.aclose()


def _attach_extension_responses(
    result: _ResponseWithSidechannelT,
    response: Any,
) -> _ResponseWithSidechannelT:
    """Populate extension_responses from the HTTP sidechannel."""
    header_obj = extract_extension_responses_header(response)
    log_extension_responses(header_obj)
    return result.model_copy(update={"extension_responses": header_obj})


# ============================================================================
# Async HTTP Facilitator Client (Default)
# ============================================================================


class HTTPFacilitatorClient(HTTPFacilitatorClientBase):
    """Async HTTP-based facilitator client.

    Communicates with remote x402 facilitator services over HTTP using
    async httpx.AsyncClient. Use with x402ResourceServer (async).

    Example:
        ```python
        from x402.http import HTTPFacilitatorClient, FacilitatorConfig

        facilitator = HTTPFacilitatorClient(FacilitatorConfig(url="https://..."))

        # In async context
        result = await facilitator.verify(payload, requirements)
        ```
    """

    def _get_sync_client(self) -> httpx.Client:
        """Get or create sync HTTP client for get_supported (initialization)."""
        import httpx

        # Create temporary sync client for initialization
        return httpx.Client(timeout=self._timeout, follow_redirects=True)

    def _get_async_client(self) -> httpx.AsyncClient:
        """Get or create async HTTP client."""
        if self._http_client is None:
            import httpx

            self._http_client = httpx.AsyncClient(timeout=self._timeout, follow_redirects=True)
        return self._http_client

    async def aclose(self) -> None:
        """Close async HTTP client if we own it."""
        if self._owns_client and self._http_client:
            await self._http_client.aclose()
            self._http_client = None

    async def __aenter__(self) -> HTTPFacilitatorClient:
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.aclose()

    # =========================================================================
    # FacilitatorClient Implementation (Async)
    # =========================================================================

    async def verify(
        self,
        payload: PaymentPayload | PaymentPayloadV1,
        requirements: PaymentRequirements | PaymentRequirementsV1,
    ) -> VerifyResponse:
        """Verify a payment with the facilitator (async).

        Args:
            payload: Payment payload to verify.
            requirements: Requirements to verify against.

        Returns:
            VerifyResponse.

        Raises:
            httpx.HTTPError: If request fails.
            ValueError: If response is invalid.
        """
        return await self._verify_http(
            payload.x402_version,
            payload.model_dump(by_alias=True, exclude_none=True),
            requirements.model_dump(by_alias=True, exclude_none=True),
        )

    async def settle(
        self,
        payload: PaymentPayload | PaymentPayloadV1,
        requirements: PaymentRequirements | PaymentRequirementsV1,
    ) -> SettleResponse:
        """Settle a payment with the facilitator (async).

        Args:
            payload: Payment payload to settle.
            requirements: Requirements for settlement.

        Returns:
            SettleResponse.

        Raises:
            httpx.HTTPError: If request fails.
            ValueError: If response is invalid.
        """
        return await self._settle_http(
            payload.x402_version,
            payload.model_dump(by_alias=True, exclude_none=True),
            requirements.model_dump(by_alias=True, exclude_none=True),
        )

    def get_supported(self) -> SupportedResponse:
        """Get supported payment kinds and extensions.

        Note: This is sync because it's called during initialization.

        Returns:
            SupportedResponse.

        Raises:
            httpx.HTTPError: If request fails.
        """
        # Use sync client for initialization (called from sync initialize())
        with self._get_sync_client() as client:
            request = client.build_request(
                "GET",
                f"{self._url}/supported",
                headers=self._get_supported_headers(),
            )
            response = client.send(request, stream=True)
            body = _read_sync_response_body(response)

            if response.status_code != 200:
                raise ValueError(
                    f"Facilitator get_supported failed ({response.status_code}): {_decode_error_body(body)}"
                )

            return _parse_facilitator_response(body, SupportedResponse, "supported")

    # =========================================================================
    # Bytes-Based Methods (Network Boundary)
    # =========================================================================

    async def verify_from_bytes(
        self,
        payload_bytes: bytes,
        requirements_bytes: bytes,
    ) -> VerifyResponse:
        """Verify payment from raw JSON bytes.

        Operates at network boundary - detects version from bytes.

        Args:
            payload_bytes: JSON bytes of payment payload.
            requirements_bytes: JSON bytes of requirements.

        Returns:
            VerifyResponse.
        """
        from ..schemas.helpers import detect_version

        version = detect_version(payload_bytes)
        payload_dict = json.loads(payload_bytes)
        requirements_dict = json.loads(requirements_bytes)

        return await self._verify_http(version, payload_dict, requirements_dict)

    async def settle_from_bytes(
        self,
        payload_bytes: bytes,
        requirements_bytes: bytes,
    ) -> SettleResponse:
        """Settle payment from raw JSON bytes.

        Operates at network boundary - detects version from bytes.

        Args:
            payload_bytes: JSON bytes of payment payload.
            requirements_bytes: JSON bytes of requirements.

        Returns:
            SettleResponse.
        """
        from ..schemas.helpers import detect_version

        version = detect_version(payload_bytes)
        payload_dict = json.loads(payload_bytes)
        requirements_dict = json.loads(requirements_bytes)

        return await self._settle_http(version, payload_dict, requirements_dict)

    # =========================================================================
    # Internal HTTP Methods (Async)
    # =========================================================================

    async def _verify_http(
        self,
        version: int,
        payload_dict: dict[str, Any],
        requirements_dict: dict[str, Any],
    ) -> VerifyResponse:
        """Internal verify via HTTP (async)."""
        client = self._get_async_client()
        request_body = self._build_request_body(version, payload_dict, requirements_dict)
        request = client.build_request(
            "POST",
            f"{self._url}/verify",
            headers=self._get_verify_headers(),
            json=request_body,
        )
        response = await client.send(request, stream=True)
        body = await _read_async_response_body(response)

        if response.status_code != 200:
            raise ValueError(
                f"Facilitator verify failed ({response.status_code}): {_decode_error_body(body)}"
            )

        result = _parse_facilitator_response(body, VerifyResponse, "verify")
        return _attach_extension_responses(result, response)

    async def _settle_http(
        self,
        version: int,
        payload_dict: dict[str, Any],
        requirements_dict: dict[str, Any],
    ) -> SettleResponse:
        """Internal settle via HTTP (async)."""
        client = self._get_async_client()
        request_body = self._build_request_body(version, payload_dict, requirements_dict)
        request = client.build_request(
            "POST",
            f"{self._url}/settle",
            headers=self._get_settle_headers(),
            json=request_body,
        )
        response = await client.send(request, stream=True)
        body = await _read_async_response_body(response)

        if response.status_code != 200:
            raise ValueError(
                f"Facilitator settle failed ({response.status_code}): {_decode_error_body(body)}"
            )

        result = _parse_facilitator_response(body, SettleResponse, "settle")
        return _attach_extension_responses(result, response)


# ============================================================================
# Sync HTTP Facilitator Client
# ============================================================================


class HTTPFacilitatorClientSync(HTTPFacilitatorClientBase):
    """Sync HTTP-based facilitator client.

    Communicates with remote x402 facilitator services over HTTP using
    sync httpx.Client. Use with x402ResourceServerSync (sync).

    Example:
        ```python
        from x402.http import HTTPFacilitatorClientSync, FacilitatorConfig

        facilitator = HTTPFacilitatorClientSync(FacilitatorConfig(url="https://..."))

        # Sync usage
        result = facilitator.verify(payload, requirements)
        ```
    """

    def _get_client(self) -> httpx.Client:
        """Get or create HTTP client."""
        if self._http_client is None:
            import httpx

            self._http_client = httpx.Client(timeout=self._timeout, follow_redirects=True)
        return self._http_client

    def close(self) -> None:
        """Close HTTP client if we own it."""
        if self._owns_client and self._http_client:
            self._http_client.close()
            self._http_client = None

    def __enter__(self) -> HTTPFacilitatorClientSync:
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()

    # =========================================================================
    # FacilitatorClientSync Implementation
    # =========================================================================

    def verify(
        self,
        payload: PaymentPayload | PaymentPayloadV1,
        requirements: PaymentRequirements | PaymentRequirementsV1,
    ) -> VerifyResponse:
        """Verify a payment with the facilitator.

        Args:
            payload: Payment payload to verify.
            requirements: Requirements to verify against.

        Returns:
            VerifyResponse.

        Raises:
            httpx.HTTPError: If request fails.
            ValueError: If response is invalid.
        """
        return self._verify_http(
            payload.x402_version,
            payload.model_dump(by_alias=True, exclude_none=True),
            requirements.model_dump(by_alias=True, exclude_none=True),
        )

    def settle(
        self,
        payload: PaymentPayload | PaymentPayloadV1,
        requirements: PaymentRequirements | PaymentRequirementsV1,
    ) -> SettleResponse:
        """Settle a payment with the facilitator.

        Args:
            payload: Payment payload to settle.
            requirements: Requirements for settlement.

        Returns:
            SettleResponse.

        Raises:
            httpx.HTTPError: If request fails.
            ValueError: If response is invalid.
        """
        return self._settle_http(
            payload.x402_version,
            payload.model_dump(by_alias=True, exclude_none=True),
            requirements.model_dump(by_alias=True, exclude_none=True),
        )

    def get_supported(self) -> SupportedResponse:
        """Get supported payment kinds and extensions.

        Returns:
            SupportedResponse.

        Raises:
            httpx.HTTPError: If request fails.
        """
        client = self._get_client()
        request = client.build_request(
            "GET",
            f"{self._url}/supported",
            headers=self._get_supported_headers(),
        )
        response = client.send(request, stream=True)
        body = _read_sync_response_body(response)

        if response.status_code != 200:
            raise ValueError(
                f"Facilitator get_supported failed ({response.status_code}): {_decode_error_body(body)}"
            )

        return _parse_facilitator_response(body, SupportedResponse, "supported")

    # =========================================================================
    # Bytes-Based Methods (Network Boundary)
    # =========================================================================

    def verify_from_bytes(
        self,
        payload_bytes: bytes,
        requirements_bytes: bytes,
    ) -> VerifyResponse:
        """Verify payment from raw JSON bytes.

        Operates at network boundary - detects version from bytes.

        Args:
            payload_bytes: JSON bytes of payment payload.
            requirements_bytes: JSON bytes of requirements.

        Returns:
            VerifyResponse.
        """
        from ..schemas.helpers import detect_version

        version = detect_version(payload_bytes)
        payload_dict = json.loads(payload_bytes)
        requirements_dict = json.loads(requirements_bytes)

        return self._verify_http(version, payload_dict, requirements_dict)

    def settle_from_bytes(
        self,
        payload_bytes: bytes,
        requirements_bytes: bytes,
    ) -> SettleResponse:
        """Settle payment from raw JSON bytes.

        Operates at network boundary - detects version from bytes.

        Args:
            payload_bytes: JSON bytes of payment payload.
            requirements_bytes: JSON bytes of requirements.

        Returns:
            SettleResponse.
        """
        from ..schemas.helpers import detect_version

        version = detect_version(payload_bytes)
        payload_dict = json.loads(payload_bytes)
        requirements_dict = json.loads(requirements_bytes)

        return self._settle_http(version, payload_dict, requirements_dict)

    # =========================================================================
    # Internal HTTP Methods
    # =========================================================================

    def _verify_http(
        self,
        version: int,
        payload_dict: dict[str, Any],
        requirements_dict: dict[str, Any],
    ) -> VerifyResponse:
        """Internal verify via HTTP."""
        client = self._get_client()
        request_body = self._build_request_body(version, payload_dict, requirements_dict)
        request = client.build_request(
            "POST",
            f"{self._url}/verify",
            headers=self._get_verify_headers(),
            json=request_body,
        )
        response = client.send(request, stream=True)
        body = _read_sync_response_body(response)

        if response.status_code != 200:
            raise ValueError(
                f"Facilitator verify failed ({response.status_code}): {_decode_error_body(body)}"
            )

        result = _parse_facilitator_response(body, VerifyResponse, "verify")
        return _attach_extension_responses(result, response)

    def _settle_http(
        self,
        version: int,
        payload_dict: dict[str, Any],
        requirements_dict: dict[str, Any],
    ) -> SettleResponse:
        """Internal settle via HTTP."""
        client = self._get_client()
        request_body = self._build_request_body(version, payload_dict, requirements_dict)
        request = client.build_request(
            "POST",
            f"{self._url}/settle",
            headers=self._get_settle_headers(),
            json=request_body,
        )
        response = client.send(request, stream=True)
        body = _read_sync_response_body(response)

        if response.status_code != 200:
            raise ValueError(
                f"Facilitator settle failed ({response.status_code}): {_decode_error_body(body)}"
            )

        result = _parse_facilitator_response(body, SettleResponse, "settle")
        return _attach_extension_responses(result, response)

File: ruby/x402/lib/x402/http/facilitator_client.rb
# frozen_string_literal: true

require 'faraday'
require 'json'
require_relative '../schemas/base'
require_relative '../schemas/payments'
require_relative '../schemas/responses'

module X402
  module HTTP
    ##
    # HTTP-based facilitator client using Faraday.
    #
    # Communicates with remote x402 facilitator services over HTTP.
    #
    # @example Basic usage
    #   client = X402::HTTP::FacilitatorClient.new(url: 'https://x402.org/facilitator')
    #   supported = client.get_supported
    #   result = client.verify(payload, requirements)
    #
    # @example With custom timeout
    #   client = X402::HTTP::FacilitatorClient.new(
    #     url: 'https://x402.org/facilitator',
    #     timeout: 60
    #   )
    #
    # @example With authentication
    #   client = X402::HTTP::FacilitatorClient.new(
    #     url: 'https://x402.org/facilitator',
    #     auth_headers: {
    #       'Authorization' => 'Bearer token123'
    #     }
    #   )
    class FacilitatorClient
      # Default facilitator URL
      DEFAULT_URL = 'https://x402.org/facilitator'

      ##
      # @return [String] facilitator base URL
      attr_reader :url

      ##
      # @return [String] facilitator identifier
      attr_reader :identifier

      ##
      # Create a new HTTP facilitator client.
      #
      # @param url [String] facilitator base URL
      # @param timeout [Numeric] request timeout in seconds (default: 30)
      # @param auth_headers [Hash, nil] optional authentication headers
      # @param identifier [String, nil] optional identifier (defaults to URL)
      # @param connection [Faraday::Connection, nil] optional custom Faraday connection
      def initialize(url: DEFAULT_URL, timeout: 30, auth_headers: nil, identifier: nil, connection: nil)
        @url = url.chomp('/')
        @timeout = timeout
        @auth_headers = auth_headers || {}
        @identifier = identifier || @url
        @connection = connection
      end

      ##
      # Get supported payment kinds and extensions.
      #
      # @return [SupportedResponse] supported kinds, extensions, and signers
      # @raise [HTTPError] if request fails
      #
      # @example
      #   supported = client.get_supported
      #   supported.kinds.each { |kind| puts "#{kind.scheme} on #{kind.network}" }
      def get_supported
        response = connection.get('supported') do |req|
          merge_auth_headers(req, @auth_headers)
        end

        handle_response(response) do |body|
          SupportedResponse.from_json(JSON.generate(body))
        end
      end

      ##
      # Verify a payment with the facilitator.
      #
      # @param payload [PaymentPayload] payment payload to verify
      # @param requirements [PaymentRequirements] requirements to verify against
      # @return [VerifyResponse] verification result
      # @raise [HTTPError] if request fails
      #
      # @example
      #   result = client.verify(payload, requirements)
      #   puts result.valid? ? "Valid" : "Invalid: #{result.invalid_reason}"
      def verify(payload, requirements)
        request_body = build_request_body(
          payload.x402_version,
          JSON.parse(payload.to_json),
          JSON.parse(requirements.to_json)
        )

        response = connection.post('verify') do |req|
          req.headers['Content-Type'] = 'application/json'
          merge_auth_headers(req, @auth_headers)
          req.body = JSON.generate(request_body)
        end

        handle_response(response) do |body|
          VerifyResponse.from_json(JSON.generate(body))
        end
      end

      ##
      # Settle a payment with the facilitator.
      #
      # @param payload [PaymentPayload] payment payload to settle
      # @param requirements [PaymentRequirements] requirements for settlement
      # @return [SettleResponse] settlement result
      # @raise [HTTPError] if request fails
      #
      # @example
      #   result = client.settle(payload, requirements)
      #   puts "Transaction: #{result.transaction}" if result.success?
      def settle(payload, requirements)
        request_body = build_request_body(
          payload.x402_version,
          JSON.parse(payload.to_json),
          JSON.parse(requirements.to_json)
        )

        response = connection.post('settle') do |req|
          req.headers['Content-Type'] = 'application/json'
          merge_auth_headers(req, @auth_headers)
          req.body = JSON.generate(request_body)
        end

        handle_response(response) do |body|
          SettleResponse.from_json(JSON.generate(body))
        end
      end

      ##
      # Verify payment from raw JSON bytes.
      #
      # @param payload_bytes [String] JSON string of payment payload
      # @param requirements_bytes [String] JSON string of requirements
      # @return [VerifyResponse] verification result
      # @raise [HTTPError] if request fails
      def verify_from_bytes(payload_bytes, requirements_bytes)
        version = Helpers.detect_version(payload_bytes)
        payload_dict = JSON.parse(payload_bytes)
        requirements_dict = JSON.parse(requirements_bytes)

        request_body = build_request_body(version, payload_dict, requirements_dict)

        response = connection.post('verify') do |req|
          req.headers['Content-Type'] = 'application/json'
          merge_auth_headers(req, @auth_headers)
          req.body = JSON.generate(request_body)
        end

        handle_response(response) do |body|
          VerifyResponse.from_json(JSON.generate(body))
        end
      end

      ##
      # Settle payment from raw JSON bytes.
      #
      # @param payload_bytes [String] JSON string of payment payload
      # @param requirements_bytes [String] JSON string of requirements
      # @return [SettleResponse] settlement result
      # @raise [HTTPError] if request fails
      def settle_from_bytes(payload_bytes, requirements_bytes)
        version = Helpers.detect_version(payload_bytes)
        payload_dict = JSON.parse(payload_bytes)
        requirements_dict = JSON.parse(requirements_bytes)

        request_body = build_request_body(version, payload_dict, requirements_dict)

        response = connection.post('settle') do |req|
          req.headers['Content-Type'] = 'application/json'
          merge_auth_headers(req, @auth_headers)
          req.body = JSON.generate(request_body)
        end

        handle_response(response) do |body|
          SettleResponse.from_json(JSON.generate(body))
        end
      end

      private

      ##
      # Get or create Faraday connection.
      #
      # @return [Faraday::Connection] connection instance
      def connection
        @connection ||= Faraday.new(url: @url) do |f|
          f.request :json
          f.response :json, content_type: /\bjson$/
          f.adapter Faraday.default_adapter
          f.options.timeout = @timeout
          f.options.open_timeout = @timeout
        end
      end

      ##
      # Build request body for verify/settle endpoints.
      #
      # @param version [Integer] protocol version
      # @param payload_dict [Hash] payment payload as hash
      # @param requirements_dict [Hash] requirements as hash
      # @return [Hash] request body
      def build_request_body(version, payload_dict, requirements_dict)
        {
          x402Version: version,
          paymentPayload: to_json_safe(payload_dict),
          paymentRequirements: to_json_safe(requirements_dict)
        }
      end

      ##
      # Convert object to JSON-safe format (handles large integers).
      #
      # @param obj [Object] object to convert
      # @return [Object] JSON-safe object
      def to_json_safe(obj)
        # Ruby handles large integers natively, but ensure consistency
        JSON.parse(JSON.generate(obj))
      end

      ##
      # Handle HTTP response and parse result.
      #
      # @param response [Faraday::Response] HTTP response
      # @yield [body] block to parse response body
      # @yieldparam body [Hash] parsed response body
      # @return [Object] parsed result
      # @raise [HTTPError] if response status is not 2xx
      def handle_response(response)
        unless response.success?
          raise HTTPError.new(
            response.status,
            response.body.is_a?(Hash) ? response.body : response.body.to_s
          )
        end

        yield response.body
      end

      ##
      # Merge authentication headers into request.
      #
      # @param req [Faraday::Request] request object
      # @param headers [Hash] headers to merge
      def merge_auth_headers(req, headers)
        headers.each do |key, value|
          req.headers[key] = value
        end
      end
    end
  end
end
